Setup issues

Hi!

I'm having issues setting up.
I first tried the binaries found here:
https://prometheus.io/download/#graphite_exporter

Seemed to be working fine, until I tried to send some test data to it.

Startup:
```bash
root@MONIT-01:/opt/prometheus/graphite_exporter_0# ./graphite_exporter 
level=info ts=2020-02-16T10:02:30.856Z caller=main.go:260 msg="Starting graphite_exporter" version_info="(version=0.7.0-rc1, branch=HEAD, revision=b11c8a66bb01034aacadff5bcc6bf073b7803340)"
level=info ts=2020-02-16T10:02:30.857Z caller=main.go:261 build_context="(go=go1.13.6, user=root@5b94ccd60948, date=20200121-10:31:27)"
level=info ts=2020-02-16T10:02:30.859Z caller=main.go:341 msg="Listening on :9108"
```

Then, from another SSH session:
```bash
root@MONIT-01:~# echo "test_tcp 1234 $(date +%s)" | nc localhost 9109
root@MONIT-01:~# 
```

I immediately got pinged by the first session, which had now printed the following error before quitting:
```bash
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x8a7df9]

goroutine 26 [running]:
github.com/prometheus/statsd_exporter/pkg/mapper.(*MetricMapper).GetMapping(0xc000136180, 0xc0000c8400, 0x8, 0x9d0091, 0x5, 0x0, 0x0, 0x0)
        /app/vendor/github.com/prometheus/statsd_exporter/pkg/mapper/mapper.go:240 +0xc9
main.(*graphiteCollector).processLine(0xc000100280, 0xc0000c8400, 0x18)
        /app/main.go:137 +0x478
main.(*graphiteCollector).processLines(0xc000100280)
        /app/main.go:123 +0x5f
created by main.newGraphiteCollector
        /app/main.go:107 +0x150
```

After fiddling for awhile (mostly reinstalling graphite_exporter and go) I thought maybe the issue was with the precompiled binary.

So I attempted to compile from source, with no luck.
This is the error I got (and keep getting): http://paste.ubuntu.com/p/NZNhdkzBkD/

Any ideas?

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,3 +1,7 @@
+
+## Unrealesed
+* [BUGFIX] Fix bug around statsd mapper cache init ([#118](https://github.com/prometheus/graphite_exporter/pull/118))
+
 ## 0.7.0-rc1 / 2020-01-21
 
 * [CHANGE] Update logging library and flags ([#109](https://github.com/prometheus/graphite_exporter/pull/109))

--- a/e2e/e2e_test.go
+++ b/e2e/e2e_test.go
@@ -108,3 +108,84 @@ rspamd.spam_count 3 NOW`
 		}
 	}
 }
+
+func TestIssue111(t *testing.T) {
+	cwd, err := os.Getwd()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	webAddr, graphiteAddr := fmt.Sprintf("127.0.0.1:%d", 9108), fmt.Sprintf("127.0.0.1:%d", 9109)
+	exporter := exec.Command(
+		filepath.Join(cwd, "..", "graphite_exporter"),
+		"--web.listen-address", webAddr,
+		"--graphite.listen-address", graphiteAddr,
+	)
+	err = exporter.Start()
+	if err != nil {
+		t.Fatalf("execution error: %v", err)
+	}
+	defer exporter.Process.Kill()
+
+	for i := 0; i < 10; i++ {
+		if i > 0 {
+			time.Sleep(1 * time.Second)
+		}
+		resp, err := http.Get("http://" + webAddr)
+		if err != nil {
+			continue
+		}
+		defer resp.Body.Close()
+		if resp.StatusCode == http.StatusOK {
+			break
+		}
+	}
+
+	now := time.Now()
+
+	input := `rspamd.actions.add_header 2 NOW
+rspamd.actions.greylist 0 NOW
+rspamd.actions.no_action 24 NOW
+rspamd.actions.reject 1 NOW
+rspamd.actions.rewrite_subject 0 NOW
+rspamd.actions.soft_reject 0 NOW
+rspamd.bytes_allocated 4165268944 NOW
+rspamd.chunks_allocated 4294966730 NOW
+rspamd.chunks_freed 0 NOW
+rspamd.chunks_oversized 1 NOW
+rspamd.connections 1 NOW
+rspamd.control_connections 1 NOW
+rspamd.ham_count 24 NOW
+rspamd.learned 2 NOW
+rspamd.pools_allocated 59 NOW
+rspamd.pools_freed 171 NOW
+rspamd.scanned 27 NOW
+rspamd.shared_chunks_allocated 34 NOW
+rspamd.spam_count 3 NOW`
+	input = strings.NewReplacer("NOW", fmt.Sprintf("%d", now.Unix())).Replace(input)
+
+	conn, err := net.Dial("tcp", graphiteAddr)
+	if err != nil {
+		t.Fatalf("connection error: %v", err)
+	}
+	defer conn.Close()
+	_, err = conn.Write([]byte(input))
+	if err != nil {
+		t.Fatalf("write error: %v", err)
+	}
+
+	resp, err := http.Get("http://" + path.Join(webAddr, "metrics"))
+	if err != nil {
+		t.Fatalf("get error: %v", err)
+	}
+	defer resp.Body.Close()
+	b, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatalf("read error: %v", err)
+	}
+	for _, s := range []string{"rspamd_actions_add_header 2", "rspamd_connections 1"} {
+		if !strings.Contains(string(b), s) {
+			t.Fatalf("Expected %q in %q – input: %q – time: %s", s, string(b), input, now)
+		}
+	}
+}

--- a/main.go
+++ b/main.go
@@ -82,6 +82,7 @@ func (s graphiteSample) String() string {
 type metricMapper interface {
 	GetMapping(string, mapper.MetricType) (*mapper.MetricMapping, prometheus.Labels, bool)
 	InitFromFile(string, int) error
+	InitCache(cacheSize int)
 }
 
 type graphiteCollector struct {
@@ -271,6 +272,8 @@ func main() {
 			level.Error(logger).Log("msg", "Error loading metric mapping config", "err", err)
 			os.Exit(1)
 		}
+	} else {
+		c.mapper.InitCache(0)
 	}
 
 	if *dumpFSMPath != "" {

--- a/main_test.go
+++ b/main_test.go
@@ -41,7 +41,9 @@ func (m *mockMapper) GetMapping(metricName string, metricType mapper.MetricType)
 func (m *mockMapper) InitFromFile(string, int) error {
 	return nil
 }
+func (m *mockMapper) InitCache(int) {
 
+}
 func TestProcessLine(t *testing.T) {
 
 	type testCase struct {

